File: Diameter of Binary Tree/Leetcode_543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.height = 0
        def dfs(node: TreeNode, h: int) -> int:
            if not node: return 0
            l, r = dfs(node.left, h+1), dfs(node.right, h+1)
            self.height = max(self.height, l+r)
            return 1 + max(l, r)
        dfs(root, 0)
        return self.height

File: Diameter of Binary Tree/Leetcode_543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// woefully slow 
class Solution {
private:
  int dfs(TreeNode* node) {
        return !node ? 0: max(dfs(node->left), dfs(node->right)) + 1;
    };    
public:
    int diameterOfBinaryTree(TreeNode* root) {
        return !root ? 0 : max(dfs(root->left) + dfs(root->right), 
                max(diameterOfBinaryTree(root->left),  diameterOfBinaryTree(root->right)));
    }

};
